### PR TITLE
[Merged by Bors] - feat: update the `ContDiff` definition to integrate analytic functions in the hierarchy

### PR DESCRIPTION
**Pull request by sgouezel**
We modify the definition of `ContDiff` so that the smoothness exponent belongs to `WithTop (ℕ∞)`, where top smoothness (denoted with `ω` in the `ContDiff` scope) means analyticity (together with analyticity of all the iterated derivatives).

This will make it possible to write theorems like: consider a function which is `C^2` over the reals or complexes, or analytic over a general field. Then... Many theorems in calculus hold under these assumptions (and it's the way things are written in Bourbaki), so having this possibility is a necessary improvement unless one wants to duplicate everything.

The PR is mostly not changing the API: theorems that were proved for `C^n` functions with `n ∈ ℕ or n = ∞` are now proved for `n ∈ ℕ or n = ∞ or n = ω` (unless they are not true for analytic functions, but it turns out in most cases things are fine). Of course, results have to be added in the files `ContDiff.Def` and `ContDiff.Basic` to deal with analytic functions.

I know that the PR is huge and hard to review. I have tried to separate all the other results in other PRs, but once one changes the basic definition one has to fix everything downstream to get mathlib compiling, so I can not make it shorter...

---

- [x] depends on: #15270
- [x] depends on: #16720
- [x] depends on: #16725
- [x] depends on: #16752 
- [x] depends on: #16753 
- [x] depends on: #16843
- [x] depends on: #16844
- [x] depends on: #16891
- [x] depends on: #16972
- [x] depends on: #16985 
- [x] depends on: #17002
- [x] depends on: #17146
- [x] depends on: #17164 
- [x] depends on: #17169 
- [x] depends on: #17170
- [x] depends on: #17218 
- [x] depends on: #17221 
- [x] depends on: #17400
- [x] depends on: #18722 
- [x] depends on: #18723 
- [x] depends on: #19374 
- [x] depends on: #19442 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/Calculus/ContDiff/FaaDiBruno.lean [hunk 1]
```diff
@@ -0,0 +1,964 @@
...
+one can form a continuous multilinear map in `n`
+variables by applying `p m` to each part of the partition, and then
+applying `f` to the resulting vector. It is called `c.compAlongOrderedFinpartition f p`. -/
+def compAlongOrderedFinpartition
+    (f : ContinuousMultilinearMap 𝕜 (fun _i : Fin c.length ↦ F) G)
+    (p : ∀ (i : Fin c.length), E[×c.partSize i]→L[𝕜] F) :
+    E[×n]→L[𝕜] G where
+  toFun v := f (c.applyOrderedFinpartition p v)
+  map_add' v i x y := by
+    cases Subsingleton.elim ‹_› (instDecidableEqFin _)
```
Comment (b-mehta) [inline Mathlib/Analysis/Calculus/ContDiff/FaaDiBruno.lean:747]: Not entirely related to the rest of this PR, but I think this is symptomatic of a very annoying diamond in `DecidableEq (Fin _)` that I've also experienced :(

Comment (sgouezel) [reply Mathlib/Analysis/Calculus/ContDiff/FaaDiBruno.lean:747]: Not really, it's more the definition of multilinear maps, which don't assume `DecidableEq` but instead quantify over all such instances in the fields `map_add'` and `map_smul'`. So, on a concrete type with a concrete `DecidableEq` instance, you always have to make this small dance when proving that something is multilinear.

Comment (b-mehta) [reply Mathlib/Analysis/Calculus/ContDiff/FaaDiBruno.lean:747]: Ah, I see, okay!

File: Mathlib/Analysis/Calculus/ContDiff/FiniteDimension.lean [hunk 2]
```diff
@@ -28,8 +28,11 @@ section FiniteDimensional
 
 open Function Module
 
+open scoped ContDiff
+
 variable [CompleteSpace 𝕜]
 
+
```
Comment (urkud) [inline Mathlib/Analysis/Calculus/ContDiff/FiniteDimension.lean:35]:
```suggestion
```

File: Mathlib/Analysis/Complex/CauchyIntegral.lean [hunk 3]
```diff
@@ -579,7 +579,7 @@ theorem _root_.DifferentiableOn.analyticOn {s : Set ℂ} {f : ℂ → E} (hd : D
 differentiable on `s`. -/
 protected theorem _root_.DifferentiableOn.contDiffOn {s : Set ℂ} {f : ℂ → E} {n : ℕ}
     (hd : DifferentiableOn ℂ f s) (hs : IsOpen s) : ContDiffOn ℂ n f s :=
-  (hd.analyticOnNhd hs).contDiffOn
+  (hd.analyticOnNhd hs).contDiffOn hs.uniqueDiffOn
```
Comment (urkud) [inline Mathlib/Analysis/Complex/CauchyIntegral.lean:582]: Why `AnalyticOnNhd.contDiffOn` needs `UniqueDiffOn` now?

Comment (sgouezel) [reply Mathlib/Analysis/Complex/CauchyIntegral.lean:582]: The assumptions in `AnalyticOnNhd.contDiffOn` has changed: in the previous version, it was requiring a complete space, but no unique differentiability. While the new version drops completeness, but requires unique differentiability.

Comment (urkud) [reply Mathlib/Analysis/Complex/CauchyIntegral.lean:582]: Does it make sense to have the old version too? Anyway, could you please mention API changes like this one (i.e., not just `ENat -> WithTop ENat`) in the commit message? Also, am I right that we can upgrade this lemma from `n : Nat` to `n : WithTop ENat` for free?

Comment (sgouezel) [reply Mathlib/Analysis/Complex/CauchyIntegral.lean]: Sure, I've added back the old version, under the name `AnalyticOnNhd.contDiffOn_of_completeSpace` (and I've used it in `CauchyIntegral.lean`).

I think this lemma already use `n : WithTop ENat`.

File: Mathlib/Analysis/Complex/CauchyIntegral.lean [hunk 4]
```diff
@@ -579,7 +579,7 @@ theorem _root_.DifferentiableOn.analyticOn {s : Set ℂ} {f : ℂ → E} (hd : D
 differentiable on `s`. -/
 protected theorem _root_.DifferentiableOn.contDiffOn {s : Set ℂ} {f : ℂ → E} {n : ℕ}
```
Comment (urkud) [inline Mathlib/Analysis/Complex/CauchyIntegral.lean:580]: Will it work like this?
```suggestion
protected theorem _root_.DifferentiableOn.contDiffOn {s : Set ℂ} {f : ℂ → E} {n : WithTop ℕ +}
```

**Comment by urkud** [comment]
Otherwise LGTM. Thanks!
bors d+

**Comment by sgouezel** [comment]
bors r+
Thanks a lot for the review!

**Pull request closed without merging**